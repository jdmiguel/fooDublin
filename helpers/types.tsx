import { ReactNode } from 'react';

export type InitialAppState = {
  favorites: Restaurant[];
  relatedRestaurants: Restaurant[];
  breadcrumbs: BreadcrumbsData[];
};

// Core / ui components

export enum LogoSize {
  BIG = 'big',
  SMALL = 'small',
}

export enum LoaderType {
  CIRCLE = 'circle',
  LINE = 'line',
}

export enum LoaderMode {
  DARK = 'dark',
  LIGHT = 'light',
}

export type ListItem = {
  iconSrc?: string;
  id: number;
  name: string;
  path: string;
};

export enum CardType {
  SUGGESTION = 'suggestion',
  STANDART = 'standart',
  HIGHLIGHT = 'highlight',
}

export enum CustomLinkSize {
  BIG = 'big',
  SMALL = 'small',
}

// Restaurant data

export type Timming = {
  id: string;
  day: string;
  schedule: string;
};

export enum EntityType {
  CITY = 'city',
  SUBZONE = 'subzone',
}

export type RestaurantsRequestParams = {
  entity_id: number | null;
  cuisines: number | null;
  entity_type: EntityType.CITY | EntityType.SUBZONE;
  start?: number;
  sort?: string;
  order?: string;
  q?: string;
  count?: number;
};

export type Restaurant = {
  id: number;
  imgSrc: string;
  title: string;
  content: ReactNode | string;
  route: string;
  asRoute: string;
};

export type RawRestaurant = {
  restaurant: {
    R: {
      has_menu_status: {
        delivery: number;
        takeaway: number;
      };
      res_id: number;
      is_grocery_store: boolean;
    };
    apikey: string;
    id: number;
    name: string;
    url: string;
    location: {
      address: string;
      locality: string;
      city: string;
      city_id: number;
      latitude: number;
      longitude: number;
      zipcode: string;
      country_id: number;
      locality_verbose: string;
    };
    switch_to_order_menu: number;
    cuisines: string;
    timings: string;
    average_cost_for_two: number;
    price_range: number;
    currency: string;
    highlights: string[];
    offers: any[];
    opentable_support: number;
    is_zomato_book_res: number;
    mezzo_provider: string;
    is_book_form_web_view: number;
    book_form_web_view_url: string;
    book_again_url: string;
    thumb: string;
    user_rating: {
      aggregate_rating: number;
      rating_text: string;
      rating_color: string;
      rating_obj: {
        title: {
          text: number;
        };
        bg_color: {
          type: string;
          tint: number;
        };
      };
      votes: number;
    };
    all_reviews_count: number;
    photos_url: string;
    photo_count: number;
    menu_url: string;
    featured_image: string;
    has_online_delivery: number;
    is_delivering_now: number;
    store_type: string;
    include_bogo_offers: boolean;
    deeplink: string;
    is_table_reservation_supported: number;
    has_table_booking: number;
    events_url: string;
    phone_numbers: string;
    all_reviews: {
      reviews: [
        {
          review: any[];
        },
        {
          review: any[];
        },
        {
          review: any[];
        },
        {
          review: any[];
        },
        {
          review: any[];
        },
      ];
    };
    establishment: string[];
  };
};

export type RestaurantDetail = {
  imgSrc: string;
  thumbSrc: string;
  name: string;
  location: string;
  cuisines: string;
  timings: string;
  rating: number;
  votes: number;
  average: string;
  establishment: string;
  highlights: string[];
  phone: string;
  address: string;
};

export type Review = {
  id: number;
  userName: string;
  userImgSrc: string;
  rating: number;
  date: string;
  text: string;
};

export type RawReview = {
  review: {
    rating: number;
    review_text: string;
    id: number;
    rating_color: string;
    review_time_friendly: string;
    rating_text: string;
    timestamp: number;
    likes: number;
    user: {
      name: string;
      foodie_level: string;
      foodie_level_num: number;
      foodie_color: string;
      profile_url: string;
      profile_image: string;
      profile_deeplink: string;
    };
    comments_count: number;
  };
};

// Breadcrumbs

export enum BreadcrumbsType {
  HOME = 'home',
  SEARCH = 'search',
  DETAIL = 'detail',
  FAVORITES = 'favorite',
}

export type BreadcrumbsData = {
  text: string;
  route: string;
  asRoute: string;
  type:
    | BreadcrumbsType.HOME
    | BreadcrumbsType.SEARCH
    | BreadcrumbsType.DETAIL
    | BreadcrumbsType.FAVORITES;
};

// Actions

type AddFavoriteAction = {
  type: 'ADD_FAVORITE';
  favorite: Restaurant;
};

type DeleteFavoriteAction = {
  type: 'DELETE_FAVORITE';
  id: number;
};

type SetRelatedRestaurantsAction = {
  type: 'SET_RELATED_RESTAURANTS';
  relatedRestaurants: Restaurant[];
};

type ClearRelatedRestaurantsAction = {
  type: 'CLEAR_RELATED_RESTAURANTS';
};

type AddBreadcrumbsAction = {
  type: 'ADD_BREADCRUMBS';
  breadcrumbs: BreadcrumbsData;
};

type ReplaceBreadcrumbsAction = {
  type: 'REPLACE_BREADCRUMBS';
  index: number;
  breadcrumbs: BreadcrumbsData;
};

export type Actions =
  | AddFavoriteAction
  | DeleteFavoriteAction
  | SetRelatedRestaurantsAction
  | ClearRelatedRestaurantsAction
  | AddBreadcrumbsAction
  | ReplaceBreadcrumbsAction;
