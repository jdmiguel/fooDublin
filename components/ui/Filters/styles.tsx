import styled from 'styled-components';

export const StyledFilters = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-gap: 10px;
  @media only screen and (min-width: 768px) {
    grid-template-columns: repeat(auto-fit, minmax(150px, 1fr));
  }
`;
