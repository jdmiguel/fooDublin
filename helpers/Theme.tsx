import React from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

export const theme: DefaultTheme = {
  palette: {
    PRIMARY: '#EC7E00',
    PRIMARY_DARK: '#D48C24',
    PRIMARY_MEDIUM: '#FBA52B',
    PRIMARY_LIGHT: '#FDF3E5',
    SECONDARY: '#660000',
    LIGHT_MAX: '#FFFFFF',
    LIGHT_MEDIUM: '#F3F3F3',
    LIGHT_MIN: '#E5EAED',
    LIGHT_SOFT: '#D8D8D8',
    DARK_MAX: '#222C3C',
    DARK_MEDIUM: '#585858',
    DARK_MIN: '#89959B',
    DARK_SOFT: '#A7A7A7',
  },
};

export const renderWithTheme = (component: any) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);
