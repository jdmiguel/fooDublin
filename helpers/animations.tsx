import { keyframes, css } from 'styled-components';

const fadeIn = (endValue: number) => keyframes`
  0% { opacity: 0 }
  100% { opacity: ${endValue} }
`;

const fadeInFromPosY = (initialPosY: number) => keyframes`
  0% { opacity: 0; transform: translateY(${initialPosY}px); }
  100% { opacity: 1; ; transform: translateY(0); }
`;

const rotate = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

const moveX = keyframes`
  0% { transform: translateX(-100%); }
  100% { transform: translateX(0); }
`;

export const fadeInAnimation = css<{ endValue?: number }>`
  animation: ${({ endValue }) => fadeIn(endValue || 1)} 0.15s linear;
`;

export const fadeInFromPosYAnimation = css<{
  initialPosY?: number;
  animationDuration?: number;
  animationDelay?: number;
}>`
  animation-name: ${({ initialPosY }) => fadeInFromPosY(initialPosY || 50)};
  animation-duration: ${({ animationDuration }) => animationDuration || 0.15}s;
  animation-delay: ${({ animationDelay }) => animationDelay || 0.25}s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
`;

export const circleLoaderAnimation = css`
  animation: ${rotate} 0.8s linear infinite;
`;

export const lineLoaderAnimation = css`
  animation: ${moveX} 1s ease-in-out infinite forwards;
`;
