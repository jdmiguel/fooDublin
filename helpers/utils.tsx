import { createStore, Reducer } from 'redux';
import appReducer from '@/store/redux/reducer';
import { Restaurant } from '@/components/pages/types';

// GENERAL UTILS

type ComposableStringFunction = (text: string) => string;

export const compose =
  (...fns: ComposableStringFunction[]) =>
  (value: string) =>
    fns.reduce((acc, fn) => fn(acc), value);

export const getRandomInt = (min: number, max: number) =>
  Math.floor(Math.random() * (max - min)) + min;

//const checkAlphanumeric = (name: string) => !!name.match(/^[0-9a-zA-Z]+$/);

export const getAlphanumericText = (text: string) => text.replace(/[^a-z0-9]/gi, '');

export const getLoweredText = (text: string) => text.toLowerCase();

export const getFormattedUrlText = (text: string, isPath = false) => {
  const concatenatorSymbol = isPath ? '-' : '+';
  return text.split(' ').reduce((acc: string, next: string) => {
    const concatenator = acc && getAlphanumericText(next) ? concatenatorSymbol : '';
    const formattedText = compose(getLoweredText, getAlphanumericText);

    return `${acc}${concatenator}${formattedText(next)}`;
  }, '');
};

// RELATED RESTAURANTS

export const getRandomNumbersList = (
  excludedValue: number,
  minRandomValue: number,
  maxRandomValue: number,
) => {
  const indexArray: number[] = [];

  while (indexArray.length < 3) {
    const random = getRandomInt(minRandomValue, maxRandomValue);
    if (random !== excludedValue && !indexArray.includes(random)) {
      indexArray.push(random);
    }
  }

  return indexArray;
};

export const getCurrentRelatedRestaurants = (
  restaurants: Restaurant[],
  currentRestaurantId: number,
) => {
  const currentSuggestionIndex = restaurants.findIndex(
    (restaurants) => restaurants.id === currentRestaurantId,
  );
  const getRelatedRestaurantsIndexList = getRandomNumbersList(
    currentSuggestionIndex,
    0,
    restaurants.length,
  );

  return getRelatedRestaurantsIndexList.map(
    (relatedRestaurantsIndex) => restaurants[relatedRestaurantsIndex],
  );
};

// SEARCH AND FAVORITES PAGE TITLE

export const getTitleText = (total: number) => ({
  totalText: total > 0 ? total : 'There are no',
  restaurantText: `restaurant${total === 0 || total >= 2 ? 's' : ''}`,
});

// CreateMockStore

export const createTestStore = () => createStore(appReducer as Reducer);

type GetSSRResult<TProps> = { props: TProps } | { redirect: any } | { notFound: true };

type GetSSRFn<TProps> = (args: any) => Promise<GetSSRResult<TProps>>;

// TYPES

export type inferSSRProps<TFn extends GetSSRFn<any>> = TFn extends GetSSRFn<infer TProps>
  ? NonNullable<TProps>
  : never;

type GetStaticResult<TProps> = { props: TProps } | { redirect: any } | { notFound: true };

type GetStaticFn<TProps> = (args: any) => Promise<GetStaticResult<TProps>>;

export type inferStaticProps<TFn extends GetStaticFn<any>> = TFn extends GetStaticFn<infer TProps>
  ? NonNullable<TProps>
  : never;
