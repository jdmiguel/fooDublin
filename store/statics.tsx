import { BreadcrumbsType } from '@/components/core/types';

export const BASE_API = 'https://api.yelp.com/v3/';

export const DUBLIN_ID = 91;

export const DEFAULT_TEXT_LOADING = 'Coming right up...';

export const SEARCH_TERM_QUERY_PARAM = 'restaurants';
export const SEARCH_RADIUS_QUERY_PARAM = 1000;

export const MAX_RESTAURANT_DISPLAYED = 30;

export const MAX_SMALL_DEVICE_WIDTH = 640;

export const SCROLL_FACTOR = 1.1;
export const SCROLL_DELAY = 100;

export const SHOWING_SCROLLUP_BUTTON_HEIGHT = 900;

export const MAX_SEARCH_BREADCRUMBS = 2;

export const MAX_MOBILE_WIDTH = 768;

export enum PlaceholderText {
  BLURRED = 'Search for locals...',
  FOCUSED = 'Start typing to search...',
}

export const HIGHLIGHT_GENERIC_SRC = '/images/highlight-generic.png';
export const THUMB_GENERIC_SRC = '/images/generic-thumb.png';
export const DETAIL_GENERIC_SRC = '/images/generic-detail.png';
export const USER_GENERIC_SRC = '/images/user-avatar.svg';
export const DEFAULT_RATING_STAR_LIST = [
  { id: 1, type: 'star_outline' },
  { id: 2, type: 'star_outline' },
  { id: 3, type: 'star_outline' },
  { id: 4, type: 'star_outline' },
  { id: 5, type: 'star_outline' },
];

export enum FilterSort {
  COST = 'cost',
  RANK = 'rating',
}
export enum FilterOrder {
  ASC = 'asc',
  DESC = 'desc',
}

export const FILTERS = [
  {
    isActive: false,
    primaryText: 'cost',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.COST,
    order: FilterOrder.ASC,
    id: 1,
  },
  {
    isActive: false,
    primaryText: 'cost',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.COST,
    order: FilterOrder.DESC,
    id: 2,
  },
  {
    isActive: false,
    primaryText: 'rank',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.RANK,
    order: FilterOrder.ASC,
    id: 3,
  },
  {
    isActive: false,
    primaryText: 'rank',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.RANK,
    order: FilterOrder.DESC,
    id: 4,
  },
];

export const DEFAULT_BREADCRUMB = {
  text: 'Home',
  route: '/',
  asRoute: '/',
  type: BreadcrumbsType.HOME,
};

export const DUBLIN_COORDINATES = {
  latitude: 53.3434,
  longitude: -6.26761,
};

export const WEEK_DAYS = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

export const MAX_PRICE_PERCENT = 80;
export const MAX_EURO_PRICE_AMOUNT = 4;
