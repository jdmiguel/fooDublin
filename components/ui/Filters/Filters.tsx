import { useReducer, Dispatch } from 'react';
import { Filter } from '../../core/Filter/Filter';
import { FilterData } from '../../core/types';
import { StyledFilters } from './styles';

type FilterProps = {
  className?: string;
  data: FilterData[];
  onClick: (sort: string, order: string) => void;
};

type DataAction = { type: 'select'; id: number } | { type: 'clear' };

const dataReducer = (data: FilterData[], action: DataAction) => {
  switch (action.type) {
    case 'select':
      return data.map((dataItem) => ({
        ...dataItem,
        isActive: action.id === dataItem.id,
      }));
    case 'clear':
      return data.map((dataItem) => ({
        ...dataItem,
        isActive: false,
      }));
  }
};

export const Filters: React.FC<FilterProps> = ({ data, onClick }) => {
  const dataWithIsActiveProp = data.map((dataItem) => ({
    ...dataItem,
    isActive: false,
  }));
  const [dataState, dispatch]: [FilterData[], Dispatch<DataAction>] = useReducer(
    dataReducer,
    dataWithIsActiveProp,
  );

  const handleClick = (isActive: boolean, id: number, sort: string, order: string) => {
    dispatch(!isActive ? { type: 'select', id } : { type: 'clear' });
    onClick(!isActive ? sort : '', !isActive ? order : '');
  };

  return (
    <StyledFilters data-testid="filters">
      {dataState.map((item) => (
        <Filter
          key={item.id}
          data={item}
          onClick={() => {
            handleClick(item.isActive, item.id, item.sort, item.order);
          }}
        />
      ))}
    </StyledFilters>
  );
};
