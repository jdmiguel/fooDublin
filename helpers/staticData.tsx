import { CDN_URL_STATIC_DIRECTORY } from './utils';

export const DUBLIN_ID = 91;
export const BASE_URL = 'https://developers.zomato.com/api/v2.1/';

export const DEFAULT_TEXT_LOADING = 'Coming right up...';

export const MAX_RESTAURANT_DISPLAYED = 20;
export const MAX_RESTAURANT_RETRIEVED = 100;

export const MAX_SMALL_DEVICE_WIDTH = 640;

export const SCROLL_FACTOR = 4;
export const SCROLL_INITIAL_MOBILE_FACTOR = 10;
export const SCROLL_OFFSET_MOBILE_FACTOR = 4;
export const SCROLL_OFFSET_DESKTOP_FACTOR = 1.2;
export const SCROLL_DELAY = 4;

export const SHOWING_SCROLLUP_BUTTON_HEIGHT = 900;

export enum PlaceholderText {
  BLURRED = 'Search for locals...',
  FOCUSED = 'Start typing to search...',
}

export const DEFAULT_SUGGESTIONS = [
  {
    id: 0,
    imgSrc: '',
    firstText: '',
    secondText: '',
  },
];

export const LOCATIONS = [
  {
    id: 91,
    name: 'Dublin',
    path: 'dublin',
  },
  {
    id: 162239,
    name: 'South City West',
    path: 'south-city-west',
  },
  {
    id: 96201,
    name: 'Temple Bar',
    path: 'temple-bar',
  },
  {
    id: 95051,
    name: 'Clondalkin',
    path: 'clondalkin',
  },
  {
    id: 96181,
    name: 'Swords',
    path: 'swords',
  },
  {
    id: 96191,
    name: 'Tallaght',
    path: 'tallaght',
  },
  {
    id: 94881,
    name: 'Blanchardstown',
    path: 'blanchardstown',
  },
  {
    id: 162240,
    name: 'South City East',
    path: 'south-city-east',
  },
  {
    id: 95311,
    name: 'Dundrum',
    path: 'dundrum',
  },
  {
    id: 95341,
    name: 'Finglas',
    path: 'finglas',
  },
  {
    id: 94791,
    name: 'Ballyfermot',
    path: 'ballyfermot',
  },
  {
    id: 96081,
    name: 'Santry',
    path: 'santry',
  },
  {
    id: 95711,
    name: 'Lucan',
    path: 'lucan',
  },
  {
    id: 94671,
    name: 'Artane',
    path: 'artane',
  },
  {
    id: 95721,
    name: 'Malahide',
    path: 'malahide',
  },
  {
    id: 95301,
    name: 'Dun Laoghaire',
    path: 'dun-laoghaire',
  },
  {
    id: 96271,
    name: 'Walkinstown',
    path: 'walkinstown',
  },
  {
    id: 94871,
    name: 'Blackrock',
    path: 'blackrock',
  },
  {
    id: 95851,
    name: 'Palmerstown',
    path: 'palmerstown',
  },
  {
    id: 95981,
    name: 'Rathmines',
    path: 'rathmines',
  },
  {
    id: 94701,
    name: 'Balbriggan',
    path: 'balbriggan',
  },
  {
    id: 94761,
    name: 'Ballybrack',
    path: 'ballybrack',
  },
  {
    id: 96111,
    name: 'Smithfield',
    path: 'smithfield',
  },
  {
    id: 95181,
    name: 'Crumlin',
    path: 'crumlin',
  },
  {
    id: 94971,
    name: 'Castleknock',
    path: 'castleknock',
  },
  {
    id: 95511,
    name: 'Howth',
    path: 'howth',
  },
  {
    id: 94951,
    name: 'Cabra',
    path: 'cabra',
  },
  {
    id: 96291,
    name: 'Whitehall',
    path: 'whitehall',
  },
  {
    id: 95901,
    name: 'Portmarnock',
    path: 'portmarnock',
  },
  {
    id: 95641,
    name: 'Kimmage',
    path: 'kimmage',
  },
];

export const CUISINES = [
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/stew.svg`,
    id: 135,
    name: 'Irish',
    path: 'irish',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/pizza.svg`,
    id: 55,
    name: 'Italian',
    path: 'italian',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/bruschetta.svg`,
    id: 70,
    name: 'Mediterranean',
    path: 'mediterranean',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/rice.svg`,
    id: 3,
    name: 'Asian',
    path: 'asian',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/healthy_food.svg`,
    id: 143,
    name: 'Healthy Food',
    path: 'healthy-food',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/gulab.svg`,
    id: 148,
    name: 'Indian',
    path: 'indian',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/taco.svg`,
    id: 73,
    name: 'Mexican',
    path: 'mexican',
  },
  {
    iconSrc: `${CDN_URL_STATIC_DIRECTORY}/images/fries.svg`,
    id: 40,
    name: 'Fast Food',
    path: 'fast-food',
  },
];

export const HIGHLIGHTED_RESTAURANTS = [
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/cleaver-east.jpg`,
    id: 9100233,
    path: 'cleaver-east',
    name: 'Cleaver East Restaurant',
    description:
      'Situated in the heart of Ireland’s capital city, on East Essex Street in the popular Temple Bar area, Cleaver East Restaurant offers the finest Irish and European cuisine.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/tapas-de-lola.jpg`,
    id: 16518539,
    path: 'tapas-de-lola',
    name: 'Tapas de Lola',
    description:
      'This restaurant is a love letter to all things Spanish and flavorful. Its owners make regular pilgrimages to Spain for inspiration and pack it all into a lively menu of small plates.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/ely-wine.jpg`,
    id: 16517305,
    path: 'ely-wine',
    name: 'Ely Wine Bar',
    description:
      'Eric and his wife Michelle are the founders of this clubby place where you can eat and drink well both downstairs in the cozy basement and upstairs in an elegant dining room.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/oxmantown.jpg`,
    id: 16520229,
    path: 'oxmantown',
    name: 'Oxmantown',
    description:
      'In this small place, the art of a brilliant sandwich means every element is taken seriously, like the breakfast sandwich that comes with butcher Jack McCarthy’s black pudding.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/chapter-one.jpg`,
    id: 9100702,
    path: 'chapter-one',
    name: 'Chapter One',
    description:
      'Dublin writers are more closely associated with booze rather than food, but Chapter One is a Michelin-starred restaurant in the basement of the Dublin Writers Museum.',
  },
  {
    imgSrc: `${CDN_URL_STATIC_DIRECTORY}/images/rosa-madre.jpg`,
    id: 9101166,
    path: 'rosa-madre',
    name: 'Rosa Madre',
    description:
      'In the menu-hawking clamor of Temple Bar you might long for a quieter restaurant where you won’t feel like a tourist on a conveyor belt. If so, Rosa Madre is your place.',
  },
];

export const HIGHLIGHT_GENERIC_SRC = `${CDN_URL_STATIC_DIRECTORY}/images/highlight-generic.png`;
export const THUMB_GENERIC_SRC = `${CDN_URL_STATIC_DIRECTORY}/images/generic-thumb.png`;
export const DETAIL_GENERIC_SRC = `${CDN_URL_STATIC_DIRECTORY}/images/generic-detail.png`;
export const DEFAULT_RATING_STAR_LIST = [
  { id: 1, type: 'star_outline' },
  { id: 2, type: 'star_outline' },
  { id: 3, type: 'star_outline' },
  { id: 4, type: 'star_outline' },
  { id: 5, type: 'star_outline' },
];

export enum FilterSort {
  COST = 'cost',
  RANK = 'rating',
}
export enum FilterOrder {
  ASC = 'asc',
  DESC = 'desc',
}

export const FILTER_DATA = [
  {
    primaryText: 'cost',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.COST,
    order: FilterOrder.ASC,
    id: 1,
  },
  {
    primaryText: 'cost',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.COST,
    order: FilterOrder.DESC,
    id: 2,
  },
  {
    primaryText: 'rank',
    secondaryText: '- high to low',
    icon: 'keyboard_arrow_up',
    sort: FilterSort.RANK,
    order: FilterOrder.ASC,
    id: 3,
  },
  {
    primaryText: 'rank',
    secondaryText: '- low to high',
    icon: 'keyboard_arrow_down',
    sort: FilterSort.RANK,
    order: FilterOrder.DESC,
    id: 4,
  },
];
