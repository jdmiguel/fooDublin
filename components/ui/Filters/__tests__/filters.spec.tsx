/**
 * @jest-environment jsdom
 */

import { render, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { Filters } from '../Filters';
import { renderWithTheme } from '../../../../helpers/Theme';
import { FILTERS } from '../../../../store/statics';

describe('Component: Filter', () => {
  it('should render', () => {
    const { container } = render(renderWithTheme(<Filters onClick={() => {}} data={FILTERS} />));

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should active filter and call function on click any filter', () => {
    const handleClick = jest.fn();

    const { getByTestId } = render(
      renderWithTheme(<Filters onClick={handleClick} data={FILTERS} />),
    );

    const filters = getByTestId('filters');
    const firstFilterItem = filters.querySelector('button');
    const secondFilterItem = filters.querySelector('button:nth-of-type(2)');
    const thirdFilterItem = filters.querySelector('button:nth-of-type(3)');
    const fourthFilterItem = filters.querySelector('button:nth-of-type(4)');

    // check if background color of filterItems are the expected ones
    expect(firstFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(secondFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(thirdFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(fourthFilterItem).toHaveStyleRule('background-color', '#F3F3F3');

    fireEvent.click(firstFilterItem);

    // check if background color of firstFilterItem has changed and callback
    // function has been called
    expect(firstFilterItem).toHaveStyleRule('background-color', '#FDF3E5');
    expect(secondFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(thirdFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(fourthFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(handleClick).toHaveBeenCalled();

    fireEvent.click(firstFilterItem);

    // check if background color of firstFilterItem has changed to his defaul
    // value and callback function has been called again
    expect(firstFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(secondFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(thirdFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(fourthFilterItem).toHaveStyleRule('background-color', '#F3F3F3');
    expect(handleClick).toHaveBeenCalled();
  });
});
