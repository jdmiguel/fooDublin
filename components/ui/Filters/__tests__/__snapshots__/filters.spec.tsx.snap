// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Filter should render 1`] = `
.c1 {
  padding: 15px 8px;
  font-size: 0.9rem;
  border: 1px solid #D8D8D8;
  border-radius: 4px;
  color: #FBA52B;
  outline: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  background-color: #F3F3F3;
}

.c2 {
  text-transform: uppercase;
  font-weight: 600;
}

.c3 {
  display: none;
}

.c4 {
  font-size: 1.1rem;
  line-height: 18px;
  margin-left: 10px;
  display: block;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
}

.c0 button {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

@media only screen and (min-width:768px) {
  .c1 {
    width: auto;
    margin: 0;
    padding: 15px;
    font-size: 1rem;
  }
}

@media only screen and (min-width:992px) {
  .c1 {
    padding: 20px;
  }
}

@media only screen and (min-width:1200px) {
  .c1:hover {
    background-color: #FDF3E5;
    color: #EC7E00;
  }
}

@media only screen and (min-width:920px) {
  .c3 {
    display: block;
  }
}

@media only screen and (min-width:920px) {
  .c4 {
    display: none;
  }
}

<div
  class="c0"
  data-testid="filters"
>
  <button
    class="c1"
  >
    <span
      class="c2"
    >
      cost
    </span>
    <span
      class="c3"
    >
       - high to low
    </span>
    <i
      class="c4 material-icons"
    >
      keyboard_arrow_up
    </i>
  </button>
  <button
    class="c1"
  >
    <span
      class="c2"
    >
      cost
    </span>
    <span
      class="c3"
    >
       - low to high
    </span>
    <i
      class="c4 material-icons"
    >
      keyboard_arrow_down
    </i>
  </button>
  <button
    class="c1"
  >
    <span
      class="c2"
    >
      rank
    </span>
    <span
      class="c3"
    >
       - high to low
    </span>
    <i
      class="c4 material-icons"
    >
      keyboard_arrow_up
    </i>
  </button>
  <button
    class="c1"
  >
    <span
      class="c2"
    >
      rank
    </span>
    <span
      class="c3"
    >
       - low to high
    </span>
    <i
      class="c4 material-icons"
    >
      keyboard_arrow_down
    </i>
  </button>
</div>
`;
